src/cart/crt_swim.c

```diff
@@ -106,69 +100,82 @@ static void crt_swim_srv_cb(crt_rpc_t *rpc_req)
+	swim_id_t		 from_id = rpc_swim_input->src;
 	uint64_t		 max_delay = swim_ping_timeout_get() / 2;
-	uint64_t		 delay;
-	uint64_t		 hlc = crt_hlc_get();
-	int			 rc = -ENOTCONN;
+	uint32_t		 rcv_delay, snd_delay = 0;
+	int			 rc;
+	int			 i;
 
 	D_ASSERT(crt_is_service());
 
 	D_TRACE_DEBUG(DB_TRACE, rpc_req,
 		"incoming opc %#x with %zu updates %lu <= %lu\n",
 		rpc_req->cr_opc, rpc_swim_input->upds.ca_count,
-		self_id, rpc_swim_input->src);
-
-	rpc_priv = container_of(rpc_req, struct crt_rpc_priv, crp_pub);
-	delay = (hlc - rpc_priv->crp_req_hdr.cch_hlc) / NSEC_PER_MSEC;
-	csm->csm_msg_count++;
+		self_id, from_id);
 
-	if (csm->csm_hlc) {
-		csm->csm_avg_delay = (csm->csm_avg_delay + delay) / 2;
+	if (self_id == SWIM_ID_INVALID)
+		D_GOTO(out, rc = -DER_UNINIT);
 
-		D_TRACE_DEBUG(DB_TRACE, rpc_req,
-			      "%lu. delay: %lu ms, avg: %lu ms, max: %lu ms\n",
-			      csm->csm_msg_count, delay, csm->csm_avg_delay,
-			      max_delay);
-
-		if (delay > max_delay) {
-			swim_net_glitch_update(ctx, delay);
-			if (csm->csm_avg_delay > max_delay) {
-				delay = swim_ping_timeout_get() + max_delay;
-				swim_ping_timeout_set(delay);
-				D_ERROR("%lu: increase ping timeout to "
-					"%lu ms\n", ctx->sc_self, delay);
-
-				/* according protocol period requirement */
-				delay *= 3;
-				if (delay > swim_period_get()) {
-					swim_period_set(delay);
-					/* according protocol requirement */
-					delay *= 3;
-					swim_suspect_timeout_set(delay);
-				}
+	rpc_priv = container_of(rpc_req, struct crt_rpc_priv, crp_pub);
+	rcv_delay = (crt_hlc_get() - rpc_priv->crp_req_hdr.cch_hlc)
+		  / NSEC_PER_MSEC;
+
+	/* Update all piggybacked members with remote delays */
+	for (i = 0; i < rpc_swim_input->upds.ca_count; i++) {
+		struct swim_member_state *state;
+		swim_id_t id;
+
+		state = &rpc_swim_input->upds.ca_arrays[i].smu_state;
+		id = rpc_swim_input->upds.ca_arrays[i].smu_id;
+
+		D_SPIN_LOCK(&csm->csm_lock);
+		D_CIRCLEQ_FOREACH(cst, &csm->csm_head, cst_link) {
+			if (cst->cst_id == id) {
+				snd_delay = cst->cst_state.sms_delay;
+				snd_delay = snd_delay
+					  ? (snd_delay + state->sms_delay) / 2
+					  : state->sms_delay;
+				cst->cst_state.sms_delay = snd_delay;
+				break;
 			}
 		}
-	} else {
-		csm->csm_avg_delay = delay;
+		D_SPIN_UNLOCK(&csm->csm_lock);
 	}
-	csm->csm_hlc = hlc;
 
-	if (self_id != SWIM_ID_INVALID) {
-		rc = swim_parse_message(ctx, rpc_swim_input->src,
-					rpc_swim_input->upds.ca_arrays,
-					rpc_swim_input->upds.ca_count);
-		if (rc == -ESHUTDOWN)
-			swim_self_set(ctx, SWIM_ID_INVALID);
-		else if (rc)
-			D_ERROR("swim_parse_message() failed rc=%d\n", rc);
+	/* Update from_id member with remote delay */
+	D_SPIN_LOCK(&csm->csm_lock);
+	D_CIRCLEQ_FOREACH(cst, &csm->csm_head, cst_link) {
+		if (cst->cst_id == from_id) {
+			snd_delay = cst->cst_state.sms_delay;
+			snd_delay = snd_delay ? (snd_delay + rcv_delay) / 2
+					      : rcv_delay;
+			cst->cst_state.sms_delay = snd_delay;
+			break;
+		}
```

Comment:
This (and above) computations are duplicated by the emulator, aren't they? If this PR needs to be refreshed for some reason, it might be worth adding a comment to remind people to keep the emulator code (deliver_pkt?) in sync with the code here.

Reply:
Yes, simulator has the same implementation of this. Agree that all simulator logic should be the same with this. Probably the general comment about this should be in the simulator header. I will think about it. Thanks!